lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,983 @@
-  // FIRRTL circuit.  This doesn't communicate anything useful to the user other
-  // than flooding their terminal.
-  auto emitCircuitError = [&circuitOp](StringRef message = {}) {
-    return emitError(circuitOp.getLoc(), message);
-  };
-
-  // Examine the Circuit's Annotations doing work to remove Grand Central
-  // Annotations.  Ignore any unprocesssed annotations and rewrite the Circuit's
-  // Annotations with these when done.
-  bool removalError = false;
-  annotations.removeAnnotations([&](auto anno) {
-    if (!anno.isClass("sifive.enterprise.grandcentral.AugmentedBundleType"))
-      return false;
-
-    AugmentedBundleType bundle;
-    if (auto maybeBundle = decodeBundleType(anno))
-      bundle = maybeBundle.getValue();
-    else {
-      emitCircuitError(
-          "'firrtl.circuit' op contained an 'AugmentedBundleType' "
-          "Annotation which did not conform to the expected format")
-              .attachNote()
-          << "the problematic 'AugmentedBundleType' is: '" << anno.getDict()
-          << "'";
+  // Maybe get an "id" from an Annotation.  Generate error messages on the op if
+  // no "id" exists.
+  auto getID = [&](Operation *op,
+                   Annotation annotation) -> Optional<IntegerAttr> {
+    auto id = annotation.getMember<IntegerAttr>("id");
+    if (!id) {
+      op->emitOpError()
+          << "contained a malformed "
+             "'sifive.enterprise.grandcentral.AugmentedGroundType' annotation "
+             "that did not contain an 'id' field";
       removalError = true;
-      return false;
+      return None;
     }
+    return Optional(id);
+  };
 
-    for (auto elt : bundle.elements) {
-      std::pair<StringRef, StringRef> key = {bundle.defName, elt.name};
-      interfaceMap[key] = {elt.tpe, elt.description};
-      interfaceKeys.push_back(key);
+  // Maybe return the lone instance of a module.  Generate errors on the op if
+  // the module is not instantiated or is multiply instantiated.
+  auto exactlyOneInstance = [&](FModuleOp op,
+                                StringRef msg) -> Optional<InstanceOp> {
+    auto instances = getSymbolTable().getSymbolUses(op, circuitOp);
+    size_t numInstances;
+    if (!instances.hasValue())
+      numInstances = 0;
+    else
+      numInstances = std::distance(instances.getValue().begin(),
+                                   instances.getValue().end());
+    if (numInstances == 0) {
+      op->emitOpError() << "is marked as a GrandCentral '" << msg
+                        << "', but is never instantiated";
+      return None;
     }
+    if (numInstances > 1) {
+      auto diag = op->emitOpError()
+                  << "is marked as a GrandCentral '" << msg
+                  << "', but it is instantiated more than once";
+      for (auto instance : instances.getValue())
+        diag.attachNote(instance.getUser()->getLoc())
+            << "parent is instantiated here";
+      return None;
+    }
```

Comment:
You could use [hasSingleElement](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/STLExtras.h#L259) instead. 

```c++
  auto exactlyOneInstance = [&](FModuleOp op,
                                StringRef msg) -> Optional<InstanceOp> {
    auto instances = getSymbolTable().getSymbolUses(op, circuitOp);
    size_t numInstances;
    if (!instances.hasValue())
      op->emitOpError() << "is marked as a GrandCentral '" << msg
                        << "', but is never instantiated";
      return None;
    }
    if (!hasSingleElement(instances.getValue()) {
      auto diag = op->emitOpError()
                  << "is marked as a GrandCentral '" << msg
                  << "', but it is instantiated more than once";
      for (auto instance : instances.getValue())
        diag.attachNote(instance.getUser()->getLoc())
            << "parent is instantiated here";
      return None;
    }
```